File: ActorMessagingExample/src/main/scala/org/akka/essentials/hotswap/example/PingPongActor.scala
package org.akka.essentials.hotswap.example

import akka.actor.Actor

case class PING
case class PONG

class PingPongActor extends Actor {
  import context._
  var count = 0
  def receive: Receive = {
    case PING =>
      println("PING")
      count = count + 1
      Thread.sleep(100)
      self ! PONG
      become {
        case PONG =>
          println("PONG")
          count = count + 1
          Thread.sleep(100)
          self ! PING
          unbecome()
      }
      if(count > 10) context.stop(self)
  }
}

File: ActorMessagingExample/src/main/java/org/akka/essentials/hotswap/example/PingPongActor.java
package org.akka.essentials.hotswap.example;

import akka.actor.UntypedActor;
import akka.japi.Procedure;

public class PingPongActor extends UntypedActor {

	static String PING = "PING";
	static String PONG = "PONG";
	int count = 0;

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof String) {
			if (((String) message).matches(PING)) {
				System.out.println("PING");
				count += 1;
				Thread.sleep(100);
				getSelf().tell(PONG);
				getContext().become(new Procedure<Object>() {
					public void apply(Object message) {
						if (message instanceof String) {
							if (((String) message).matches(PONG)) {
								System.out.println("PONG");
								count += 1;
								try {
									Thread.sleep(100);
								} catch (InterruptedException e) {
									//
								}
								getSelf().tell(PING);
								getContext().unbecome();
							}
						}
					}
				});
				if (count > 10)
					getContext().stop(getSelf());
			}
		}
	}
}
